GitCommands/Git/GitModule.cs

```diff
@@ -3663,64 +3675,52 @@ public bool IsRunningGitProcess()
             return false;
         }
 
-        public static string UnquoteFileName(string fileName)
+        private static readonly Regex _escapedOctalCodePointRegex = new Regex(@"(\\([0-7]{3}))+", RegexOptions.Compiled);
+
+        /// <summary>
+        /// Unescapes any octal code points embedded within <paramref name="s"/>.
+        /// </summary>
+        /// <remarks>
+        /// If no portions of <paramref name="s"/> contain escaped data, then <paramref name="s"/> is returned.
+        /// </remarks>
+        /// <example>
+        /// <code>UnescapeOctalCodePoints(@"\353\221\220\353\213\244") == "두다"</code>
+        /// </example>
+        /// <param name="s">The string to unescape.</param>
+        /// <returns>The unescaped string, or <paramref name="s"/> if no escaped values were present.</returns>
+        [NotNull]
+        public static string UnescapeOctalCodePoints([CanBeNull] string s)
         {
-            if (fileName.IsNullOrWhiteSpace())
+            if (s == null)
```

Comment:
Shouldn't this be `string.IsNullOrWhiteSpace(s)`? There is no point to send an empty string to the regex below....

Reply:
`IsNullOrWhiteSpace` would incorrectly trim `"   "` to `""`. If anything, it should be `IsNullOrEmpty`.

`Regex.Replace` will deal with an empty string just fine and return it unchanged. I believed an empty string would be an uncommon input, so avoided the extra check. The performance difference is negligible, but it's my habit. If you want me to change it, let me know.

Reply:
Feeding data which we know doesn't require processing into Regex is wasteful - the call stack before it gets to check input string is quite deep (6+ deep).
Looking at the use, `" "` isn't a valid input anyway - a file name can't be an blank string, nor lines from patch that require decoding.

Let's change it to something like the following:

```
if (string.IsNullOrWhiteSpace(s))
{
    return s ?? string.Empty;
}
```

This way we don't return `null` to minimise a chance of NRE and preserve the original value otherwise.

Reply:
Pushed fixup with your change.

I think of escape/unescape methods as identity functions in the case that there's no subsequences to actually escape/unescape.

It just happens that this method is only used for file paths, but there may be future uses for other kinds of strings where leading/trailing whitespace is significant.

I find the handling of null/whitespace/empty strings in GE inconsistent and confusing. It has the inconvenience and cognitive load of `null` vs. `undefined` vs truthiness in JavaScript, but without their logic/semantics. Converting a null to a valid string seems weird. I would rather see fewer uses of `IsNullOrEmpty` and `IsNullOrWhiteSpace`, and more throwing of `ArgumentException` and friends.

I am lazy and I don't want to think about the multiple categories of value a variable can have. If I see a string `branchName` I want to know it's a valid branch name, not null, nor empty, nor whitespace. This is part of the logic behind the `ObjectId` type. If you have an instance, it's valid.

Hopefully in future C# 8 brings nullable reference types and they address a large class of these issues. If everyone on the project used R#, the annotations would provide many such safeguards today.

Reply:
This is a good point.
I am of an opinion all public API must perform validation of input parameters and throw if those not valid. It is not as clear cut with the return types though...